Signup: Add all verticals theme A/B test.

alaczek opened the pull request:
Work in progress.

Test live: https://calypso.live/?branch=add/vertical-themes-test

----------------------------------------
Currently, users coming from WordPress.com homepage are going through a two-step vertical survey followed by design-type selection as a part of the signup process. We’re offering there sets of themes for each design type (blog, page, grid). Vertical selection is not taken into account.

This test aims to deliver even more tailored theme selection in signup, by providing themes relevant to both vertical (the top category only) and design type. Themes that are not Headstart-able now, are not included in the test.

**Hypothesis**
Improvements to tailored theme selection in verticals should continue to show an increase in signup conversion (even if tailoring is incomplete) in combination with the new design-type step. Theme steps with tailored themes that show an increase in Signup conversion without a drop in retention or purchasing user conversion will be a success.

**Variations**
- original: the current default flow (25%)
- verticalThemes: the tested flow (25%)
- notTested: receives the current default flow, becomes available to other signup tests (50%)

alaczek commented:
@kwight this is what I have so far. I'd appreciate guidance in adding theme sets for different combinations of verticals/design-types, and anything I might need to add to make this test work.

kwight commented on client/signup/config/flows.js:
```diff
@@ -161,6 +161,13 @@ const flows = {
+	},
+
+	'new-vertical-site': {
+		steps: [ 'survey', 'desing-type', 'themes', 'domains', 'plans', 'survey-user' ],
```
typo for `design-type`

kwight commented:
@alaczek After discussing with @sirbrillig, we've decided on moving all of the theme logic (including vertical and design type choices) into a signup library – this is where the theme step will get its themes. I've got a start to this, but will need to add a couple more pieces before it's working.

In the meantime, you can see in 9eebd7f26d87f2be8522edd429cb6cb7be4cb49a what the library's `themes` data will look like. Each theme is its own object, with `design` being the design type it represents, and `verticals` being the survey categories we want that theme to be part of. So each theme is only one design type, but can be in many verticals. If you want, you can get started committing themes to that array while I finish up tying together the plumbing. You can find the translation of vertical codes to readable categories [here](https://github.com/Automattic/wp-calypso/blob/master/client/signup/steps/survey/verticals.js).

kwight commented:
Down the road, if the tests are successful, we can add the verticals and design type metadata to themes in the API, and grab everything from there.

kwight commented:
@alaczek OK, I've created a little [dataset of themes](https://github.com/Automattic/wp-calypso/pull/5962/commits/9cb94c1518a0ac77c0a247e017ca9d5dcbacef0e), already assigned to the design type they were in before; you just need to add the vertical/survey assignments. I still need to add some handling for the test and default themes (but I'll take care of it).

kwight commented:
@sirbrillig for a sanity check too, based on previous conversations.

kwight commented:
@sirbrillig Actually, I'm thinking that it'll be simpler/safer to pull out the theme handling into a separate PR and merge that first, before handling the AB test; what do you think?

sirbrillig commented on client/lib/signup/themes.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+export default function getThemes( vertical, designType ) {
+	let themeSet = themes.filter( filterDesignType, designType ).filter( filterVerticals, vertical );
```
This is a very clever use of `filter`! I'm ok with it, but I wonder if we can do it without using `this` (which can often add ambiguity)?

How about:
```
const filterByType = theme => theme.design === designType;
const filterByVertical = theme => includes( theme.verticals, vertical );
let themeSet = themes.filter( filterByType ).filter( filterByVertical );
```

sirbrillig commented on client/signup/steps/theme-selection/index.jsx:
```diff
@@ -59,7 +48,8 @@ module.exports = React.createClass( {
 	},
 
 	getThemes() {
-		return this.props.signupDependencies.themes || this.props.themes;
```
If you no longer want to support `props.themes`, you should remove it from the `propTypes` for this component.

sirbrillig commented on client/signup/steps/theme-selection/index.jsx:
```diff
@@ -59,7 +48,8 @@ module.exports = React.createClass( {
 	getThemes() {
-		return this.props.signupDependencies.themes || this.props.themes;
+		const dependencies = this.props.signupDependencies;
+		return dependencies.themes || getThemes( dependencies.surveyQuestion, dependencies.designType );
```
What happens here if `dependencies.themes`, `dependencies.surveyQuestion`, and `dependencies.designType` are all null?

sirbrillig replied:
Oh, I guess it would just fail the filters and you'd end up with the default themes. Ok.

sirbrillig commented:
> pull out the theme handling into a separate PR and merge that first

Agreed. That'd be nicer so that later when the test ends, it will be easier to tell what to revert.

sirbrillig commented:
Moved theme lib to #6004

kwight commented:
When testing, be sure to set `allowExistingUsers: true` in `active-tests.js`.

alaczek commented:
Should I be doing vertical assignments on this PR/branch, or in #6004?

kwight commented:
@alaczek Yep, you should do them on this branch, which is specifically the test (and vertical assignments are data we need for the test). #6004 just changes the way signup gets themes into the `themes` step, which made this PR (and future themes work) easier and a little more clear.

kwight commented:
Rebased to include the merged #6004 .

alaczek commented:
I added the vertical assignments and the additional themes that will be a part of this test.
I also updated the 'fallback' property, which I assume means the theme is one of the default theme set (since we now have three default theme sets - for blog, page, grid).

kwight commented:
@alaczek Ah, my fault – I should have been more clear about `fallback`. `fallback` is only the _default_ default themes, as in, when all else goes wrong or returns nothing, show these nine themes. They're the same thing as the themes we were showing before introducing design types, and they would be the themes that would be displayed if a signup flow had just a `themes` step (and not the survey or design type steps). So those should stay static regardless of what's going on with verticals and design types.

kwight commented:
More rebasing and fixes. Note that now:

* We always show nine themes.
* If the found themes (by design type and vertical) are more than nine, we return nine random themes from the set.
* If the found set is _less_ than nine, we go back to the design type pool and get a random collection of themes to make up the difference.

alaczek commented:
I tested latest changes and I'm getting a random mix for each design type, but it looks like the vertical assignment is not taken into account at all. For example I'm getting themes like Escutcheon and Libretto (assigned only to Books & Writing) in Family, Home & Lifestyle vertical. Is it possible I'm testing this wrong? Here are the steps I followed:
1. I checked out the `add/vertical-themes-test` branch.
2. I set `allowExistingUsers` to `true` in client/lib/abtest/active-tests.js
3. I started my local Calypso server and went to http://calypso.localhost:3000/
4. I cleared the local storage in console: `localStorage.clear();`
5. I run this in console to assign myself to the verticalThemes variation: `localStorage.setItem('ABTests','{"verticalThemes_20160614":"verticalThemes"}');`
6. I went to `/start/website/en/?ref=homepage` and started signup from there.

kwight commented:
Nope, you were testing correctly, I was just passing the wrong survey response after the last rebase. :)

alaczek commented:
The vertical assignments are working for me now! I tested all of them and everything checks out. I encountered an issue while trying to go back between steps though. I can go back between theme and design-type step, but I'm not able to go back to survey. 
When I hit Back on the design-type step, I'm being redirected like so:
* /start/website/survey/en
* /start/website/survey?
* /start/website/design-type

Not sure if this is some kind of quirk in local environment, or if this is a bug.

kwight commented:
> I can go back between theme and design-type step, but I'm not able to go back to survey. 

Hm, I can't replicate this, with the Back link nor the browser Back button; any console errors?..

kwight commented:
@alaczek I lied; I still had to add the fallback behaviour for the current production themes when not in the test. (that's done now). I've rebased, and I believe we're good to go now.

@michaeldcain or @sirbrillig for review

alaczek commented:
Shall we change the test start date? Originally I put in June 20.

michaeldcain commented:
@alaczek: we'll want to update the test start date right before we merge the PR. 

Could you update the initial PR comment to add some testing instructions or a description of expected behavior, along with maybe a quick test hypothesis/reason, for posterity? ([example](https://github.com/Automattic/wp-calypso/pull/4536#issuecomment-206434941))

sirbrillig commented on client/signup/steps/survey/index.jsx:
```diff
@@ -59,12 +60,13 @@ export default React.createClass( {
 		const stepOneClickHandler = ( event ) => {
 			event.preventDefault();
 			event.stopPropagation();
+			this.setState( { surveyFirstQuestion: vertical.value } );
```
I haven't completely grokked the changes to this file, but setting state inside a render method is a little concerning, mainly because it could lead to infinite loops (`setState` calls `render`, `render` calls `setState`). If you're just looking to record the vertical ID and pass it along, we already have that data inside `handleNextStep` (if configured to show two steps, which we are, the first question value is in `this.state.stepOne.value`).

kwight replied:
@sirbrillig Yes! I already figured out this was way more complex than it needed to be, just hadn't pushed yet.

kwight commented:
This is good to go code-wise; final look/test @michaeldcain or @sirbrillig ?

michaeldcain commented on client/lib/abtest/active-tests.js:
```diff
@@ -120,4 +120,23 @@ module.exports = {
 		defaultVariation: 'enabled',
 		allowExistingUsers: false,
 	},
+	readerRelatedPosts: {
```
This test was removed in 5c9d9f9.

michaeldcain commented:
Other than the extra test in `active-tests.js`, this LGTM.

sirbrillig commented on client/signup/steps/survey/index.jsx:
```diff
@@ -147,7 +147,7 @@ export default React.createClass( {
 			} );
 		}
-		SignupActions.submitSignupStep( { stepName: this.props.stepName }, [], { surveySiteType: this.props.surveySiteType, surveyQuestion: vertical.value } );
+		SignupActions.submitSignupStep( { stepName: this.props.stepName }, [], { surveySiteType: this.props.surveySiteType, surveyQuestion: this.state.stepOne.value } );
```
Juuuuust for safety, and in case someone decides to come in here later and change `isSurveyOneStep`, could you duplicate this line, putting the new version in the `if` block just above and the old version in `else` clause?

The reason being that if `isSurveyOneStep` is true, `this.state.stepOne` will be `null` and this line will throw a fatal error.

Really, we should just remove all the one-step specific code in this file, but until we do I'd prefer not to add the risk of any fatal errors.

kwight commented on client/signup/steps/survey/index.jsx:
```diff
@@ -141,13 +141,14 @@ export default React.createClass( {
 				category_id: value,
 				category_label: label
 			} );
+			SignupActions.submitSignupStep( { stepName: this.props.stepName }, [], { surveySiteType: this.props.surveySiteType, surveyQuestion: this.state.stepOne.value } );
```
@sirbrillig what do you think?

sirbrillig replied:
❤️

kwight commented:
@alaczek It looks like the only differences in theme selection by vertical are for the Blog design; is this right? For users in the test, they get the same theme choices for Homepage and Grid as if they weren't in the test?..

alaczek commented:
Yep, we're really short on non-portfolio grid themes and static front page themes, so the differences there are very minor. Most of the swapping occurs in Blog layout, and this is intentional.

alaczek commented:
I rebased and changed the start date for the test. Are we good to go @michaeldcain ?

michaeldcain commented:
:shipit:
